logfire/integrations/loguru.py

```diff
@@ -38,51 +48,49 @@ def fill_attributes(self, record: LogRecord) -> dict[str, Any]:
             while frame:  # pragma: no branch
                 if frame.f_code is _LOG_METHOD_CODE:
                     frame_locals = frame.f_locals
-                    if 'message' in frame_locals:
+                    if 'message' in frame_locals:  # pragma: no branch
                         attributes[ATTRIBUTES_MESSAGE_TEMPLATE_KEY] = frame_locals['message']
                     else:  # pragma: no cover
-                        _warn_inspection_failure()
+                        warn_at_user_stacklevel(
+                            'Failed to extract message template (span name) for loguru log.', LoguruInspectionFailed
+                        )
 
                     args = frame_locals.get('args')
-                    if isinstance(args, (tuple, list)):
+                    if isinstance(args, (tuple, list)):  # pragma: no branch
                         if args:
                             attributes[ATTRIBUTES_LOGGING_ARGS_KEY] = args
                     else:  # pragma: no cover
-                        _warn_inspection_failure()
-
-                    if record.exc_info:
-                        original_record = frame_locals.get('log_record')
-                        if isinstance(original_record, dict):
-                            message = original_record.get('message')  # type: ignore
-                            if isinstance(message, str) and record.msg.startswith(
-                                message + '\nTraceback (most recent call last):'
-                            ):
-                                # `record.msg` includes a traceback added by Loguru,
-                                # replace it with the original message.
-                                attributes[ATTRIBUTES_MESSAGE_KEY] = message
-                            else:  # pragma: no cover
-                                _warn_inspection_failure()
-                        else:  # pragma: no cover
-                            _warn_inspection_failure()
+                        warn_at_user_stacklevel('Failed to extract args for loguru log.', LoguruInspectionFailed)
+
+                    original_record: dict[str, Any] | None = frame_locals.get('log_record')
+                    if (
+                        isinstance(original_record, dict)
+                        and isinstance(message := original_record.get('message'), str)
+                        and message in record.msg
+                    ):  # pragma: no branch
```

Comment:
@alexmojaki Hi, This piece of code will drop the loguru's format, I'm wondering if this is intentional. Because the `replace it with the original message` only worked when the record has `exc_info`, but now it always does the replacement.

```py
# example.py

import logfire
from loguru import logger

log_format = "{level: <8} | {time:YYYY-MM-DD HH:mm:ss.SSS} | {message}"

logger.remove()
logfire.configure(
    send_to_logfire='if-token-present',
    service_name='xxx',
    console=logfire.ConsoleOptions(include_timestamps=False)  # remove timestamp prefix
)
logger.add(
    logfire.loguru_handler()['sink'],
    # sink=sys.stderr,
    format=log_format,
    level="INFO",
)

logger.info('hello world')
```
with `logfire==0.48.0`, output `INFO     | 2024-08-13 14:15:37.811 | hello world`
with `logfire==0.48.1`, output `hello world`

Reply:
Yes this is intentional. The integration is not designed for a loguru format. The message must be plain because it's also displayed in the UI:

<img width="430" alt="Screenshot 2024-08-13 at 12 10 33" src="https://github.com/user-attachments/assets/bb541d8d-725f-44ab-8ed4-b420c29477d1">

It should just be:

<img width="180" alt="Screenshot 2024-08-13 at 12 11 29" src="https://github.com/user-attachments/assets/24524656-f9a3-47e3-95fc-9fb8a674ca16">

How are you using logfire+loguru? Do you use the UI, or only the console? Does logfire log other non-loguru things to the console without timestamps?

The [structlog integration](https://docs.pydantic.dev/logfire/integrations/structlog/) has a `console_log` parameter which should probably be added to the loguru integration, so that logfire just doesn't log loguru logs to the console at all. Then timestamps don't need to be removed from other logs.

Reply:
Thank you. Currently, we use the `loguru` to print JSON logs to `console`, and then they will be collected to `aws.cloudwatch`, so we can do some filterings in `cloudwatch`, i.e., `$record.extra.xxx = xxx`.  It's structured and easy to filter. We are still experimenting with Logfire, it's a great project, thanks. So, both systems are running in parallel, and we don't want any breaking changes on `cloudwatch` yet, as there is a lot of predefined filtering in it.

```python
# example.py
import logfire
from loguru import logger

log_format = "{level: <8} | {time:YYYY-MM-DD HH:mm:ss.SSS} | {message}"

logger.remove()
logfire.configure(
    send_to_logfire='if-token-present',
    service_name='xxx',
    console=logfire.ConsoleOptions(include_timestamps=False)  # remove timestamp prefix
)
logger.add(
    logfire.loguru_handler()['sink'],
    # sink=sys.stderr,
    format=log_format,
    level="INFO",
    serialize=True,   # so it will print the logs in JSON format
)
```
with `logfire==0.48.0`, it will print the JSON logs to stderr.
```
{"text": "INFO     | 2024-08-14 09:25:58.212 | hello world", "record": {"elapsed": {"repr": "0:00:00.228754", "seconds": 0.228754}, "exception": null, "extra": {}, "file": {"name": "example.py", "path": "/Users/toddzhang/test/fastapi-test/tests/example.py"}, "function": "<module>", "level": {"icon": "ℹ️", "name": "INFO", "no": 20}, "line": 21, "message": "hello world", "module": "example", "name": "__main__", "process": {"id": 47410, "name": "MainProcess"}, "thread": {"id": 8462629568, "name": "MainThread"}, "time": {"repr": "2024-08-14 09:25:58.212465+08:00", "timestamp": 1723598758.212465}}}
```
with `logfire==0.48.1`, just plain text ```hello world```

Reply:
if there's a way, we print the JSON logs to `console/stderr` for collection by AWS, and only display `plain text` on the Logfire UI?

Reply:
Take a look at https://pydanticlogfire.slack.com/archives/C06EDRBSAH3/p1719808545882839?thread_ts=1719593610.501669&cid=C06EDRBSAH3

Reply:
I don't have permission to visit this Slack thread. But I found that I can add two sinks for `loguru`
- one is `sys.stderr`, and we can print JSON messages to the console.
- another is the `logfire.loguru_handler()['sink']`, and configure logfire with `console=False`

So, the log messages sent to logfire are pure text, and the messages printed to the console are in JSON format.

Thanks for your help.😊

Reply:
See https://docs.pydantic.dev/logfire/help/ for the link to join the slack.

Are you not creating any logfire logs/spans outside of integrating with loguru? You're not using any other integrations?